src/pressure/VCycleMG.cpp

```diff
@@ -248,171 +215,88 @@ void VCycleMG::pressure(Field &out, Field const &b, real t, bool sync) {
+            Solve(out, *field_mg_temporal_level, *field_residuum1_level, m_levels, sync);
         }
-        return;
-    }
-
-    //===================== levels going down ====================//
-    for (size_t i = 0; i < max_level; ++i) {
-        auto f_res0i = residuum0[i];
-        auto f_err1i = error1[i];
-        auto f_mg_tmpi = mg_temporal_solution[i];
-        auto f_res1i = residuum1[i];
-        auto f_res1ip = residuum1[i + 1];
-
-        auto d_res0i = residuum0[i]->data;
-        auto d_err1ip = error1[i + 1]->data;
-        auto d_res1ip = residuum1[i + 1]->data;
-
-        FieldType type_r0 = f_res0i->get_type();
-        FieldType type_r1 = residuum1[i + 1]->get_type();
-
-#pragma acc data present(res0i, err1i, err1ip, mg_tmpi, res1i, res1ip, out)
-        {
-            if (i == 0) {  // use p=out on finest grid
-                // smooth
-                Smooth(out, *f_mg_tmpi, *f_res1i, i, sync);
-
-                // calculate residuum
-                Residuum(*f_res0i, out, *f_res1i, i, sync);
-                boundary->apply_boundary(d_res0i, i, type_r0, sync);  // for residuum0 only Dirichlet BC
-            } else {
+    } else {
+        m_error1[0]->copy_data(out);
+        //===================== levels going down ====================//
+        for (size_t level = 0; level < m_levels; ++level) {
+            Field *field_residuum0_level = *(m_residuum0 + level);
+            Field *field_error1_level = *(m_error1 + level);
+            Field *field_error1_level_plus_1 = *(m_error1 + level + 1);
+            Field *field_mg_temporal_level = *(m_mg_temporal_solution + level);
+            Field *field_residuum1_level = *(m_residuum1 + level);
+            Field *field_residuum1_level_plus_1 = *(m_residuum1 + level + 1);
+
+            real *data_residuum0_level = field_residuum0_level->data;
+            real *data_residuum1_level_plus_1 = field_residuum1_level_plus_1->data;
+
+            FieldType type_r0 = field_residuum0_level->get_type();
+            FieldType type_r1 = field_residuum1_level_plus_1->get_type();
+
+#pragma acc data present(residuum0_level, error1_level, error1_level_plus_1, \
+                         mg_temporal_level, residuum1_level, residuum1_level_plus_1, \
+                         out)
+            {
                 // smooth
-                Smooth(*f_err1i, *f_mg_tmpi, *f_res1i, i, sync);
+                Smooth(*field_error1_level, *field_mg_temporal_level, *field_residuum1_level, level, sync);
 
                 // calculate residuum
-                Residuum(*f_res0i, *f_err1i, *f_res1i, i, sync);
-                boundary->apply_boundary(d_res0i, i, type_r0, sync);  // for residuum0 only Dirichlet BC
-            }
+                Residuum(*field_residuum0_level, *field_error1_level, *field_residuum1_level, level, sync);
+                boundary->apply_boundary(data_residuum0_level, level, type_r0, sync); // for m_residuum0 only Dirichlet BC
```

Comment:
two spaces between code and comment are needed.

Reply:
b167247